core/evidence.c

```diff
@@ -0,0 +1,114 @@
+/* Copyright (c) 2022 Intel Corporation
+ * Copyright (c) 2020-2022 Alibaba Cloud
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <stdlib.h>
+#include <string.h>
+#include <librats/err.h>
+#include <librats/log.h>
+#include <librats/evidence.h>
+
+size_t _size_of_evidence(const char *type)
+{
+	if (!strcmp(type, "sgx_epid")) {
+		return sizeof(attestation_verification_report_t);
+	} else if (!strcmp(type, "sgx_ecdsa")) {
+		return sizeof(ecdsa_attestation_evidence_t);
+	} else if (!strcmp(type, "tdx_ecdsa")) {
+		return sizeof(tdx_attestation_evidence_t);
+	} else if (!strcmp(type, "sgx_la")) {
+		return sizeof(la_attestation_evidence_t);
+	} else if (!strcmp(type, "sev_snp")) {
+		return sizeof(snp_attestation_evidence_t);
+	} else if (!strcmp(type, "sev")) {
+		return sizeof(sev_attestation_evidence_t);
+	} else if (!strcmp(type, "csv")) {
+		return sizeof(csv_attestation_evidence_t);
+	}
+
+	RATS_WARN("Unhandled evidence type %s\n", type);
+	return -1;
+}
+
+rats_err_t serialize_evidence(const attestation_evidence_t *evidence, uint8_t **buffer,
+			      size_t *buffer_size)
+{
+	size_t data_size = 0;
+	size_t evidence_size;
+	uint8_t *tmp = NULL;
+	uint8_t *p;
+	rats_err_t ret = RATS_ERR_NONE;
+
+	if (!evidence || !buffer_size)
+		return RATS_ERR_INVALID_PARAMETER;
+
+	data_size += sizeof(evidence->type);
+
+	// TODO: report can be shorter
+	evidence_size = _size_of_evidence(evidence->type);
+	if (evidence_size == -1)
+		return RATS_ERR_INVALID_PARAMETER;
+	data_size += evidence_size;
+
+	if (!buffer) {
+		*buffer_size = data_size;
+		return RATS_ERR_NONE;
+	}
+
+	if (!*buffer) {
+		ret = RATS_ERR_NO_MEM;
+		tmp = malloc(data_size);
+		if (!tmp)
+			goto err;
+		p = tmp;
+	} else {
+		p = *buffer;
+	}
+
+	memcpy(p, (uint8_t *)evidence, sizeof(evidence->type));
+	p += sizeof(evidence->type);
+	memcpy(p, ((uint8_t *)evidence) + offsetof(attestation_evidence_t, ecdsa), evidence_size);
+	p += evidence_size;
+
+	if (!*buffer) {
+		*buffer = tmp;
+		tmp = NULL;
+	} else {
+		*buffer = p;
+	}
+
+	*buffer_size = data_size;
+
+	ret = RATS_ERR_NONE;
+err:
+	if (tmp)
+		free(tmp);
+
+	return ret;
+}
+
+rats_err_t deserialize_evidence(attestation_evidence_t *evidence, uint8_t **buffer)
```

Comment:
Is it better to let `buffer` a `const` input parameter, and calculate the offset of `custom_claims_buffer` outside the function?